connectivity/drivers/cellular/QUECTEL/BG96/QUECTEL_BG96_CellularStack.cpp

```diff
@@ -358,29 +354,57 @@ nsapi_size_or_error_t QUECTEL_BG96_CellularStack::socket_sendto_impl(CellularSoc
 
     // Send
     if (socket->proto == NSAPI_UDP) {
+        if (size > BG96_MAX_SEND_SIZE) {
+            return NSAPI_ERROR_PARAMETER;
+        }
         char ipdot[NSAPI_IP_SIZE];
         ip2dot(address, ipdot);
         _at.cmd_start_stop("+QISEND", "=", "%d%d%s%d", socket->id, size,
                            ipdot, address.get_port());
-    } else {
-        if (socket->tls_socket) {
-            _at.cmd_start_stop("+QSSLSEND", "=", "%d%d", socket->id, size);
-        } else {
-            _at.cmd_start_stop("+QISEND", "=", "%d%d", socket->id, size);
+        _at.resp_start(">");
+        _at.write_bytes((uint8_t *)data, size);
+        _at.resp_start();
+        _at.set_stop_tag("\r\n");
+        // Possible responses are SEND OK, SEND FAIL or ERROR.
+        char response[16];
+        response[0] = '\0';
+        _at.read_string(response, sizeof(response));
+        _at.resp_stop();
+        if (strcmp(response, "SEND OK") != 0) {
+            return NSAPI_ERROR_DEVICE_ERROR;
         }
-    }
+    } else {
+        const char *buf = (const char *)data;
+        nsapi_size_t blk = BG96_MAX_SEND_SIZE;
+        nsapi_size_t count = size;
 
-    _at.resp_start(">");
-    _at.write_bytes((uint8_t *)data, size);
-    _at.resp_start();
-    _at.set_stop_tag("\r\n");
-    // Possible responses are SEND OK, SEND FAIL or ERROR.
-    char response[16];
-    response[0] = '\0';
-    _at.read_string(response, sizeof(response));
-    _at.resp_stop();
-    if (strcmp(response, "SEND OK") != 0) {
-        return NSAPI_ERROR_DEVICE_ERROR;
+        while ((count > 0)) {
```

Comment:
It would be worth handling the zero length packet case.

Reply:
This wasn't handled before either. What is the expected behaviour in this case?

It should return 0 in this case as it will skip the AT write part completely. We could do an argument validation at the beginning of the function, but I believe it is not necessary.

Reply:
Zero length packets are valid and should be _written_. That was the case before, the function was not exiting if the size in input was 0. With the change, nothing is written because the loop is not even entered if `size = 0`.
I suggest the use of a `do { } while(...);` construct instead of a `while`.

Reply:
Didn't know that, thanks. Done.

Reply:
Going to contradict @pan- here. Zero-length packets are a meaningless concept for TCP, because there are no packets. It's a stream. Writing 0 bytes is a straight no-op in a lot of implementations. Sometimes it might give you back a pending error indication, including a potential "connection reset" but a zero-length read is a more common idiom to get that, because a fair few write implementations exit early on len=0.

Zero-length packets are valid for UDP, and those should be passed to and from the underlying modem (in the hope that it will support them).

But it would be acceptable for this layer to silently bin 0-length TCP writes. Indeed, it would be the best thing to do if you knew that the modem went weird and returned an error rather than "0". But if this modem seems to work fine and return 0 for a 0-length write in the normal case, you may as well pass it through, as it now does,

Reply:
I did not test if the modem handles this case.

Regarding zero-length packets, I have read that they might be used to keep the TCP connection alive. Ref: https://osqa-ask.wireshark.org/questions/11863/why-do-tcp-clients-send-packets-with-no-data/

Reply:
Thanks @kjbracey-arm for providing more insight. I agree it doesn't makes much sense to send a zero length tcp packet if the ACK flag isn't controlled. 

I went through the application note of the module; `AT+QISEND=<connectID>,0` is supported for the _keepalive_ use case but it won't work with the current implementation as the expected return code is `OK` and not `SEND OK`. It also returns the length of sent data, acknowledged data and unacknowledged data. 

I think the best option for such use case would be to support `NSAPI_KEEPALIVE` using the `AT+QICFG="tcp/keepalive"` AT command in `setsockopt` of the module. 

@boraozgen My apologies for the confusion, I think it is best to revert back to the `while` loop instead of `do { } while()` and drop 0 length packet until a better implementation is made.

Reply:
> Regarding zero-length packets, I have read that they might be used to keep the TCP connection alive.

Certainly there can be zero-data segments under the hood in TCP, but from the API point of view, it's a stream. You can't do anything to (reliably) influence packet size via the read/write API.

As pan- says, keep-alives have to be configured via a separate control.

Reply:
Interesting to read that that modem gives "send 0" a special function. That does make sense - writing 0 in the usual sense is meaningless, so may as well assign a meaning to the command.

For the BSD/POSIX `write`/`send` API though, the "be a no-op" for 0 is required by the general "stream" semantics. It has to be a no-op for all stream-type devices - UART/TCP/files...

Reply:
> @boraozgen My apologies for the confusion, I think it is best to revert back to the `while` loop instead of `do { } while()` and drop 0 length packet until a better implementation is made.

Reverted.